smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/scheduler/SmallFileScheduler.java

```diff
@@ -0,0 +1,299 @@
+
+      // Check if container file is null
+      String containerFilePath = actionInfo.getArgs().get(
+          SmallFileCompactAction.CONTAINER_FILE);
+      if (containerFilePath == null || containerFilePath.isEmpty()) {
+        return false;
+      }
+
+      // Check if small file list is null or empty
+      String smallFiles = actionInfo.getArgs().get(HdfsAction.FILE_PATH);
+      if (smallFiles == null || smallFiles.isEmpty()) {
+        return false;
+      }
+
+      // Get valid small file list according to the small file lock map
+      ArrayList<String> smallFileList = new Gson().fromJson(
+          smallFiles, new ArrayList<String>().getClass());
+      Iterator<String> iterator = smallFileList.iterator();
+      while (iterator.hasNext()) {
+        String smallFile = iterator.next();
+        if (smallFile == null || smallFile.isEmpty()
+            || smallFilesLock.contains(smallFile)) {
+          iterator.remove();
+        } else {
+          FileState fileState = new FileState(smallFile,
+              FileState.FileType.NORMAL, FileState.FileStage.PROCESSING);
+          try {
+            metaStore.insertUpdateFileState(fileState);
+          } catch (MetaStoreException e) {
+            LOG.error("Failed to insert file state.", e);
+          }
+        }
+      }
+
+      if (smallFileList.isEmpty()) {
+        return false;
+      }
+      smallFilesLock.addAll(smallFileList);
+
+      // Update container file map, save the info indicated whether it already exists
+      try {
+        if (dfsClient.exists(containerFilePath)) {
+          containerFileMap.put(
+              actionId, new ContainerFileExistInfo(containerFilePath, true));
+        } else {
+          containerFileMap.put(
+              actionId, new ContainerFileExistInfo(containerFilePath, false));
+        }
+      } catch (IOException e) {
+        LOG.error("Failed to check if the container file is exists: "
+            + containerFilePath, e);
+        return false;
+      }
+      smallFilesMap.put(actionId, smallFileList);
+
+      // Reset args of the action
+      Map<String, String> args = new HashMap<>(2);
+      args.put(SmallFileCompactAction.CONTAINER_FILE,
+          actionInfo.getArgs().get(SmallFileCompactAction.CONTAINER_FILE));
+      args.put(SmallFileCompactAction.FILE_PATH, new Gson().toJson(smallFileList));
+      actionInfo.setArgs(args);
+      return true;
+    }
+
+    return true;
+  }
+
+  @Override
+  public ScheduleResult onSchedule(ActionInfo actionInfo, LaunchAction action) {
+    long actionId = actionInfo.getActionId();
+    if (ACTIONS.get(1).equals(actionInfo.getActionName())) {
+      String containerFilePath = containerFileMap.get(actionId).containerFilePath;
+
+      // Check if container file is locked and retry
+      if (containerFilesLock.containsKey(containerFilePath)) {
+        int retryNum = containerFilesLock.get(containerFilePath);
+        if (retryNum > MAX_RETRY_COUNT) {
+          LOG.error(
```

Comment:
Why have a maximum number of retry restriction here? Also, the restriction is implemented on the container file?